### PR TITLE
Livesplit - Provide a TimeCalc URL for Season/Trilogy speedruns

### PR DESCRIPTION
**Pull request by alex73630**
Generating and printing a TimeCalc link as required for Season/Full game runs from Speedrun.com

This will print a link like this one in the console, ready to be submitted, using Livesplit output.
[https://solderq35.github.io/fg-time-calc/?mode=0&fs3=1&ft2=1&f3t1=1&f4t0=1&d=:&o1=1&fps=&t1=49&c1=Paris&t2=106&c2=Sapienza&t3=94&c3=Marrakesh&t4=79&c4=Bangkok&t5=76&c5=Colorado&t6=45&c6=Hokkaido&t8=74&c8=Sapienza%20reset%201&t9=75&c9=Sapienza%20reset%202&t10=42&c10=Sapienza%20reset%203&t11=42&c11=Sapienza%20reset%204&t12=27&c12=Colorado%20reset%201&t13=38&c13=Hokkaido%20reset%201](https://solderq35.github.io/fg-time-calc/?mode=0&fs3=1&ft2=1&f3t1=1&f4t0=1&d=:&o1=1&fps=&t1=49&c1=Paris&t2=106&c2=Sapienza&t3=94&c3=Marrakesh&t4=79&c4=Bangkok&t5=76&c5=Colorado&t6=45&c6=Hokkaido&t8=74&c8=Sapienza%20reset%201&t9=75&c9=Sapienza%20reset%202&t10=42&c10=Sapienza%20reset%203&t11=42&c11=Sapienza%20reset%204&t12=27&c12=Colorado%20reset%201&t13=38&c13=Hokkaido%20reset%201)

### CODE REVIEW

**Review by fu5ha — CHANGES_REQUESTED** [review]
Very cool idea, I support it! Unfortunately, there's an important corner case that isn't handled yet:

If you "complete" a mission, but have red guns, or otherwise didn't complete your objective, the autosplitter allows you to just restart the level and it will unsplit. The timecalc info tracking should support this case as well. See lines 111-130 for where this is handled for livesplit itself (i.e. where the unsplit call happens). You should be able to just get the latest entry in the timeCalcEntries and set its isComplete to false, if I'm not missing anything.

File: components/types/types.ts
Comment (LennardF1989) [inline components/types/types.ts:1265]: Wondering if types.ts is the place for this. Not sure what the policy is in this regard, but this seems specific to LiveSplit and not to types from the game. Shouldn't it go/remain in the livesplit sub-folder then?

File: components/discordRp.ts
Comment (LennardF1989) [inline components/discordRp.ts:315]: The changes in this file are technically unrelated to the livesplit integration.

**Review by LennardF1989 — COMMENTED** [review]
My 2 cents.

File: components/discordRp.ts
Comment (RDIL) [reply components/discordRp.ts:315]: They are - he's using this function to get mission names and such from scene paths.

File: components/types/types.ts
Comment (alex73630) [reply components/types/types.ts:1265]: I decided to move it to a new `livesplit.ts` types file but I think it's a bit too much for only one interface, as we don't use more of those on this specific module.

Perhaps, maybe the interface could live in the module code itself?
I'd love to get more feedback on this as I don't really know how we should go with this.

Thoughts @RDIL ?

Comment (RDIL) [reply components/types/types.ts]: I don't really have a specific preference, they can go wherever tbh

File: components/livesplit/liveSplitManager.ts
Comment (fu5ha) [inline components/livesplit/liveSplitManager.ts:399]: Added return type and simplified the logic a bit. Also, fixed the fact that the returned time wasn't floored
```ts
    private _addMissionTime(time: Seconds): Seconds {
        let computedTime = Math.floor(time)
        
        // always add at least minimum, which is usually 0 except on cutscenes where
        // you can gain an advantage by restarting in cs (bangkok, sgail specific starts)
        if (time <= this._resetMinimum) {
            computedTime = this._resetMinimum
        } else if (time > 0 && time <= 1) {
            // if in game time is between 0 and 1, add full second
            computedTime = 1
        }

        this._currentMissionTotalTime += computedTime
        this._campaignTotalTime += computedTime
        return computedTime
    }
```

File: components/livesplit/liveSplitManager.ts
Comment (fu5ha) [inline components/livesplit/liveSplitManager.ts:440]: probably get rid of this log before merging

**Review by fu5ha — CHANGES_REQUESTED** [review]
I didn't really look too hard at the link generation code, but other than that looks good other than comments :)

**Review by RDIL — APPROVED** [review]
Link gen + discord RP changes LGTM, but Fusha's feedback still remains.

File: components/livesplit/liveSplitManager.ts
Comment (alex73630) [reply components/livesplit/liveSplitManager.ts]: Removed!

**Comment by alex73630** [comment]
I updated the code following @fu5ha recommendations and now waiting on approval.

Thanks everyone for the feedback!

File: components/types/types.ts
Comment (alex73630) [reply components/types/types.ts]: I'll leave it like this then. We'll probably discuss this later if we have a need to re-organize typings for whatever reason.

**Review by fu5ha — APPROVED** [review]

**Pull request merged**
